### ISSUE TITLE
GitHub Advisory Mirroring recursion does not properly handle secondary rate limits

### ISSUE THREAD

**Issue by officerNordberg** [DependencyTrack/dependency-track#4291]
### Current Behavior

GitHub Mirroring fails nearly daily and is not resilient.  While troubleshooting after weeks of ~30 failures per day due I believe to the recursive nature of the mirroring I finally added a sleep statement between all graphql calls which resulted getting the first updates to the VULNERABILITY table since the 11th, the last day it ran without incident.    It's hard to confirm this as the CREATED column is not populated on the table but UPDATED is and inferring Descending Primary Keys as "created dates" there are large gaps in GHSA primary keys that correlate to the days I received notifications of failed mirror attempts.  I think the recursion is trying to handle this but it just results in 30 or more error Notifications being sent instead of just 1. 

I added some additional details to the error message. 

```
2024-10-21 16:49:15,024 ERROR [GitHubAdvisoryMirrorTask] An error was encountered retrieving advisories with HTTP Status : 403 Forbidden 
{  "documentation_url": "https://docs.github.com/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits",  
"message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. 
If you reach out to GitHub Support for help, please include the request ID DEAD:DEED39:7ADE8B:ED56A1:67168609."}
```
curling my user I see these headers
```
< x-github-media-type: github.v4; format=json
< x-ratelimit-limit: 5000
< x-ratelimit-remaining: 4845
< x-ratelimit-reset: 1729532625
< x-ratelimit-used: 155
< x-ratelimit-resource: graphql
```

Attempted to get feedback on this issue as a discussion first.
https://github.com/DependencyTrack/dependency-track/discussions/4239

### Steps to Reproduce

1. Enable GitHub Mirroring and notifications when it fails.

### Expected Behavior

Secondary rate limits should be handled with a retry with backoff strategy. 

### Dependency-Track Version

4.11.x

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

11.22

### Browser

Google Chrome

### Checklist

- [x] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [x] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by nscuro** [comment]
@officerNordberg In #4239 you mention:

> Is everyone else using this with enterprise accounts and not personal accounts?

Regarding secondary rate limits, the [GitHub docs](https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#about-secondary-rate-limits) say:

> Make too many concurrent requests. No more than 100 concurrent requests are allowed. This limit is shared across the REST API and GraphQL API.

Could it be that your organization has other API tokens that are used for interacting with GitHub's APIs? I'm wondering if simply using a personal token would at least temporarily resolve the issue for you.

In any case, given this statement in the docs:

> These secondary rate limits are subject to change without notice. You may also encounter a secondary rate limit for undisclosed reasons.

we will need a retry mechanism no matter if the above resolves your problem for now.

**Comment by officerNordberg** [comment]
> Could it be that your organization has other API tokens that are used for interacting with GitHub's APIs? I'm wondering if simply using a personal token would at least temporarily resolve the issue for you.
> 
Nope, the tokens I've used are personal and not Enterprise. I also created new ones and purged old ones.  Nothing else was using the tokens.  I posted the rate-limit output, it was not hitting the primary limits, it was hitting the secondary limits. I asked about Enterprise tokens to see why this wasn't a more pervasive issue.  The only thing that came to mind was everyone else IS using Enterprise tokens which might be why no one else seems to be seeing this issue.  The other thought was maybe most people are experiencing mirroring issues but just not seeing it because they don't have notifications configured like ours.     

> In any case, given this statement in the docs:
> 
> > These secondary rate limits are subject to change without notice. You may also encounter a secondary rate limit for undisclosed reasons.
> 
> we will need a retry mechanism no matter if the above resolves your problem for now.

Yes, the current download the entire thing daily without any retry appears to be in need of replacement.

**Comment by Mik4sa** [comment]
I get this error too:
```txt
2024-12-16 11:12:17,480 INFO [GitHubAdvisoryMirrorTask] Starting GitHub Advisory mirroring task
[...]
2024-12-16 11:14:28,196 ERROR [GitHubAdvisoryMirrorTask] An error was encountered retrieving advisories with HTTP Status : 403 Forbidden
2024-12-16 11:14:28,197 INFO [GitHubAdvisoryMirrorTask] GitHub Advisory mirroring complete
2024-12-16 11:14:28,197 INFO [GitHubAdvisoryMirrorTask] Time spent (total): 130717ms
```

I started a fresh install of Dependecy-Track last friday and activated the `GitHub Advisories` just today.

**Comment by feberlepilz** [comment]
We have the same issue. We have changed the code to get more information when 403 is returned:
```
2024-12-18 10:55:39,464 ERROR [GitHubAdvisoryMirrorTask] MORE INFO - responseString: {  "documentation_url": "https://docs.github.com/free-pro-team@latest/rest/overview/rate-limits-for-the-rest-api#about-secondary-rate-limits",  "message": "You have exceeded a secondary rate limit. Please wait a few minutes before you try again. If you reach out to GitHub Support for help, please include the request ID BA0A:12BAFE:15A38BB:1615D09:6762AA2A."}
```
So there is a problem with the secondary rate limit of the GitHub API.

**Comment by markehack** [comment]
The problem seems to be even worse than described.  Since you do not attempt to do any incremental updates, the query hits the same limit every time and does not update the GHSA vulnerabilities. 

The limit seems to have been strictly enforced around the beginning of October 2024 since that is the last update we got. Since that date, the query just fails every day.

**Comment by nscuro** [comment]
@markehack Thanks for sharing that. Will prioritize and aim to get this into 4.12.3.

**Comment by nscuro** [comment]
Turns out open-vulnerability-clients did not have retry handling for GitHub built-in, so we had to build our own. 

The system now handles *primary* rate limiting (i.e. exceeding the allowed request quota), as well as *secondary* rate limiting (indicated by the `retry-after` response header).

Mirroring is now incremental, and we checkpoint after each processed batch of vulnerabilities, so even in failure cases we won't be starting from zero again.

I also added some basic progress logging so the status can be observed, e.g.

```
INFO [GitHubAdvisoryMirrorTask] GHSAs were not incrementally mirrored before; Mirroring all GHSAs
...
INFO [GitHubAdvisoryMirrorTask] Mirrored 19601/21144 GHSAs (92%); Last committed modification timestamp: 2024-11-13T22:38:18Z
...
INFO [GitHubAdvisoryMirrorTask] Successfully mirrored 21144 GHSAs in PT14M33.044902667S (last committed modification timestamp: 2025-01-25T00:54:52Z)
```

When encountering rate limiting, we'll log a warning:

```
WARN [GitHubAdvisoryMirrorTask] Rate limiting detected; GitHub indicates retries to be acceptable after PT1M; Will wait and try again.
```

To prevent retries from blocking worker threads for too long, I capped the maximum allowed retry delay at 3min. So if primary rate limiting kicks in, and the request quota only resets in an hour, we short-circuit gracefully.

From my subjective observation during testing, it is highly unlikely to run into primary rate limiting. The request quota given by GitHub is generous. Secondary rate limiting seems to be mostly capped to 1min flat.
